sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/CastingSyntaxSuite.scala

```diff
@@ -0,0 +1,102 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.parser;
+
+import org.apache.spark.sql.catalyst.analysis.{AnalysisTest, UnresolvedAttribute, UnresolvedFunction, UnresolvedStar}
+import org.apache.spark.sql.catalyst.expressions.{Alias, Cast, Expression, Literal}
+import org.apache.spark.sql.catalyst.parser.{CatalystSqlParser, ParseException, ParserInterface}
+import org.apache.spark.sql.types.{DoubleType, IntegerType}
+
+class CastingSyntaxSuite extends AnalysisTest {
+  import org.apache.spark.sql.catalyst.dsl.expressions._
+  val defaultParser = CatalystSqlParser
+
+  def assertEqual(
+      sqlCommand: String,
+      e: Expression,
+      parser: ParserInterface = defaultParser): Unit = {
+    compareExpressions(parser.parseExpression(sqlCommand), e)
+  }
+
+  def assertFails(sql: String, errorMsg: String): Unit = {
+    val e = intercept[ParseException](defaultParser.parseExpression(sql))
+    assert(e.getMessage.contains(errorMsg))
+  }
+
+  test("literals") {
+    assertEqual("123::double", Cast(Literal(123), DoubleType))
+    assertEqual("'123'::double", Cast(Literal("123"), DoubleType))
+    assertEqual("'123'::int", Cast(Literal("123"), IntegerType))
+    assertEqual("'123.0'::double", Cast(Literal("123.0"), DoubleType))
+    assertEqual("'123.0' :: double", Cast(Literal("123.0"), DoubleType))
+    assertEqual("`123`::double", Cast(UnresolvedAttribute(Seq("123")), DoubleType))
+
+    assertEqual("`123::double`", UnresolvedAttribute(Seq("123::double")))
+  }
+
+  test("named expressions") {
+    assertEqual("123::double as v", Alias(Cast(Literal(123), DoubleType), "v")())
+    assertEqual("123::double v", Alias(Cast(Literal(123), DoubleType), "v")())
+    assertEqual("123 :: double v", Alias(Cast(Literal(123), DoubleType), "v")())
+    assertEqual("abc::double v", Alias(Cast(UnresolvedAttribute("abc"), DoubleType), "v")())
+    assertEqual("`abc`::double v", Alias(Cast(UnresolvedAttribute("abc"), DoubleType), "v")())
+    assertEqual("abc.def::double v",
+      Alias(Cast(UnresolvedAttribute(Seq("abc", "def")), DoubleType), "v")())
+    assertEqual("`abc.def`::double v",
+      Alias(Cast(UnresolvedAttribute(Seq("abc.def")), DoubleType), "v")())
+  }
+
+  test("boolean expressions") {
+    assertEqual("(a and b) :: int", Cast('a && 'b, IntegerType))
+    assertEqual("(a or b) :: int", Cast('a || 'b, IntegerType))
+  }
+
+  test("arithmetic expressions") {
+    assertEqual("(a - b) :: int", Cast('a - 'b, IntegerType))
+    assertEqual("(a * b) :: int", Cast('a * 'b, IntegerType))
```

Comment:
can we also test `a + b::int` and make sure it produces `Add(a, Cast(b, int))`

Reply:
added.